### PR TITLE
Add basic help thread

### PR DESCRIPTION
**Pull request by digit01Wave**
#912
# Player view 
image

image

# Game control view
image

image
image

After claiming question
image
image

### CODE REVIEW

File: ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+        public PuzzleThreadsModel(PuzzleServerContext serverContext, UserManager<IdentityUser> userManager) : base(serverContext, userManager)
+        {
+        }
+
+        /// <summary>
+        /// The team of the current user. 
+        /// This value will only be filled in the play route and if the player actually has a team.
+        /// </summary>
+        public Team Team { get; set; }
```
Comment (tabascq) [inline ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs:27]: EventSpecificPageModel already has GetTeamAsync, which ensures we only ask once...

File: ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+        /// <summary>
```
Comment (tabascq) [inline ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs:29]: Use case to think about: During an event, Game Control is going to want to easily filter this down to unclaimed messages only. Also, filter to specific puzzles.

Also, ideally GC will be able to see this page update live without needing to refresh. Not something that needs to be tackled in this first PR.

File: ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            }
+            else if (EventRole != EventRole.admin)
+            {
+                throw new NotSupportedException($"EventRole [{EventRole}] is not supported in PuzzleThreads");
+            }
+
+            LatestMessagesFromEachThread = messages
+                .GroupBy(message => message.ThreadId)
+                .Select(group => group.OrderByDescending(message => message.CreatedDateTimeInUtc).First())
+                .ToList();
```
Comment (tabascq) [inline ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs:64]: ...Async

**Review by tabascq — CHANGES_REQUESTED** [review]
Great job! This is awesome!

File: ServerCore/Pages/Threads/PuzzleThread.cshtml.cs [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+                recipients.AddRange(_context.Messages
+                    .Where(message => message.ThreadId == newMessage.ThreadId && message.IsFromGameControl)
+                    .Select(message => message.Sender.Email));
+            }
+
+            if (recipients.Any())
+            {
+                MailHelper.Singleton.SendPlaintextWithoutBcc(recipients,
+                    $"{newMessage.Subject} thread update!",
```
[the remark this answers is not in the document]
Comment (digit01Wave) [reply ServerCore/Pages/Threads/PuzzleThread.cshtml.cs:273]: The subject is automatically set to
"[TeamName]PuzzleName" for team puzzles
and "[PlayerName]PuzzleName" for single player puzzles

[the remark this answers is not in the document]
Comment (tabascq) [reply ServerCore/Pages/Threads/PuzzleThread.cshtml.cs:273]: Great!

File: ServerCore/Pages/Threads/PuzzleThreads.cshtml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                <td>
+                    @(item.Text.Length > 93 ? $"{item.Text.Substring(0, 90)}..." : item.Text)
+                </td>
+                <td>
+                    @item.Sender.Name
+                </td>
+                <td>
+                    @if (!item.ClaimerID.HasValue && !item.IsFromGameControl)
+                    {
+                        <p>&#x26A0; Claim me!</p>
```
[the remark this answers is not in the document]
Comment (digit01Wave) [reply ServerCore/Pages/Threads/PuzzleThreads.cshtml:77]: I don't know if I want the button to immediately claim the message. I really just wanted the author to look at the message first to determine if they have enough time to claim the message.

File: ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs (hunk 2 above)
Comment (digit01Wave) [reply ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs:29]: I am putting all requests here
https://github.com/PuzzleServer/mainpuzzleserver/issues/969

File: ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs (hunk 3 above)
Comment (digit01Wave) [reply ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs:64]: Can you clarify this comment?

Comment (tabascq) [reply ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs:64]: ToListAsync instead of ToList and await it

Comment (digit01Wave) [reply ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs:64]: Turns out that messages is already an IEnumerable so it doesn't have a ToListAsync()
image

File: ServerCore/Pages/Submissions/Index.cshtml [hunk 6]
```diff
@@ -416,8 +416,7 @@
     @if (!Model.PuzzleState.IsEmailOnlyMode)
     {
         <h3>Need some help?</h3>
-        <p>While solving this puzzle, if you've run into a bug, think there's an error, or are not having fun anymore, <a href="@mailtoUrl">use this link to contact the author for @Model.Puzzle.PlaintextName</a>!</p>
-        <p>A pre-populated email form will appear and you should add details to the message body. Remember: the author may only know specifics about this puzzle and not others in the event.</p>
+        <p>While solving this puzzle, if you've run into a bug, think there's an error, or are not having fun anymore, <a asp-Page="/Threads/PuzzleThread" asp-route-puzzleid="@Model.Puzzle.ID" asp-route-teamId="@Model.Team?.ID" asp-route-playerId="@Model.LoggedInUser.ID">Click this link</a> to ask for help!</p>
```
[the remark this answers is not in the document]
Comment (digit01Wave) [reply ServerCore/Pages/Submissions/Index.cshtml]: done

File: ServerCore/Pages/Threads/PuzzleThread.cshtml [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    ViewData["AdminRoute"] = "/Threads/PuzzleThread";
+    ViewData["AuthorRoute"] = "/Threads/PuzzleThread";
+    ViewData["PlayRoute"] = "/Threads/PuzzleThread";
+    ViewData["RoutingPuzzleId"] = ViewContext.RouteData.Values["puzzleId"];
+}
+<script>
+    function EditText(id) {
+        let textAreaId = "TextAreaEdit_" + id;
+        let textAreaEditBox = document.getElementById(textAreaId);
+        textAreaEditBox.style.display = "block";
```
[the remark this answers is not in the document]
Comment (digit01Wave) [reply ServerCore/Pages/Threads/PuzzleThread.cshtml]: this is great! I have updated it

File: ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            else if (EventRole == EventRole.author)
+            {
+                HashSet<int> authorPuzzleIds = UserEventHelper.GetPuzzlesForAuthorAndEvent(_context, Event, LoggedInUser)
+                    .Select(puzzle => puzzle.ID)
+                    .ToHashSet();
+                messages = messages.Where(message => authorPuzzleIds.Contains(message.Puzzle.ID));
+            }
+            else if (EventRole != EventRole.admin)
+            {
+                throw new NotSupportedException($"EventRole [{EventRole}] is not supported in PuzzleThreads");
```
Comment (tabascq) [inline ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs:61]: admins should be able to see all threads in the event

File: ServerCore/Pages/Threads/PuzzleThread.cshtml [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    else
+    {
+        <a asp-page="/Threads/PuzzleThreads">Back to thread list</a>
+    }
+</div>
+
+<h4>Thread for puzzle @Model.Puzzle.Name</h4>
+<hr />
+@{
+    bool shouldMakeDark = false;
```
Comment (tabascq) [inline ServerCore/Pages/Threads/PuzzleThread.cshtml:46]: no longer needed

File: ServerCore/Pages/Threads/PuzzleThread.cshtml [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+}
+@foreach (var message in Model.Messages)
+{
+    string classString = "message-container";
+    if (shouldMakeDark) {
```
Comment (tabascq) [inline ServerCore/Pages/Threads/PuzzleThread.cshtml:51]: no longer needed

File: ServerCore/Pages/Threads/PuzzleThreads.cshtml [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    <tbody>
+        @foreach (var item in Model.LatestMessagesFromEachThread)
+        {
+            <tr>
+                <td>
+                    @if (Model.EventRole == ModelBases.EventRole.admin || Model.EventRole == ModelBases.EventRole.author)
+                    {
+                        <a asp-Page="/Puzzles/Edit" asp-route-puzzleid=@item.Puzzle.ID>@item.Puzzle.Name</a>
+                    }
+                    else if (Model.Event.EmbedPuzzles)
```
Comment (tabascq) [inline ServerCore/Pages/Threads/PuzzleThreads.cshtml:53]: these next two links are the same. Either make them different or remove the if.

File: ServerCore/Pages/Threads/PuzzleThread.cshtml [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    {
+        <a asp-page="/Threads/PuzzleThreads">Back to thread list</a>
+    }
+</div>
+
+<h4>Thread for puzzle @Model.Puzzle.Name</h4>
+<hr />
+@foreach (var message in Model.Messages)
+{
+    <div class="message-container">
```
Comment (tabascq) [inline ServerCore/Pages/Threads/PuzzleThread.cshtml:47]: You deleted the message-game-control class here, but the CSS rule still exists. Accidental deletion or unused rule?

**Review by tabascq — APPROVED** [review]

**Pull request merged**